pkgs/os-specific/linux/guvcview/default.nix

```diff
@@ -13,19 +35,38 @@ stdenv.mkDerivation rec {
     sha256 = "11byyfpkcik7wvf2qic77zjamfr2rhji97dpj1gy2fg1bvpiqf4m";
   };
 
-  buildInputs =
-    [ SDL2
-      alsaLib
-      ffmpeg
-      gtk3
-      intltool
-      libusb1
-      libv4l
-      pkgconfig
-      portaudio
-      udev
-      gsl
-    ] ++ stdenv.lib.optional pulseaudioSupport libpulseaudio;
+  nativeBuildInputs = [
+    intltool
+    pkgconfig
+  ]
+    ++ stdenv.lib.optionals (useGtk) [ wrapGAppsHook ]
+    ++ stdenv.lib.optionals (useQt) [ wrapQtAppsHook ]
+  ;
+
+  buildInputs = [
+    SDL2
+    alsaLib
+    ffmpeg
+    libusb1
+    libv4l
+    portaudio
+    udev
+    gsl
+    libpng
+    sfml
+  ] 
+    ++ stdenv.lib.optionals (pulseaudioSupport) [ libpulseaudio ]
+    ++ stdenv.lib.optionals (useGtk) [ gtk3 ]
+    ++ stdenv.lib.optionals (useQt) [
+      qt5.qtbase
```

Comment:
```suggestion
      qtbase
```

Reply:
`guvcview` doesn't use qt5 as it's `mkDerivation` so that doesn't work. `qstopmotion` uses and there `qt5.` is not used.

Reply:
@doronbehar qt5's mkDerivation doesn't put those attributes in scope. You'll just need to call `guvcview` with `libsForQt5.callPackage`.

Reply:
it also seems qstopmotion is bugged on master https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/video/qstopmotion/default.nix#L14

Reply:
@doronbehar qt5's mkDerivation doesn't put those attributes in scope. You'll just need to call guvcview with libsForQt5.callPackage.

But `guvcview` is not necessarily a qt5 application, it can be compiled as a GTK program so how could I incorporate that choice with this approach?

Reply:
> it also seems qstopmotion is bugged on master https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/video/qstopmotion/default.nix#L14

So you want qstopmotion to be called with `libsForQt5` and then that `with qt5;` won't be needed?

Reply:
@doronbehar It's already called with it so you just have to remove `with qt5` and add those all as args.

Reply:
> > @doronbehar qt5's mkDerivation doesn't put those attributes in scope. You'll just need to call guvcview with libsForQt5.callPackage.
> 
> But `guvcview` is not necessarily a qt5 application, it can be compiled as a GTK program so how could I incorporate that choice with this approach?

The callPackage function doesn't actually bring any extra dependencies. It just brings those attributes into scope.

Reply:
> The callPackage function doesn't actually bring any extra dependencies. It just brings those attributes into scope. (and ensures correct versions)

I understand but I don't understand how would you expect (in guvcview) that `stdenv.mkDerivation` will be used if `useGtk == true` vs `libsForQt5.mkDerivation` if `useQt == true`. I guess that's possible using a `let ... in` where the `stdenv` will be chosen but I don't understand what would be the benefit considering the price of readability and that only 1 `qt5` package is used there.

Reply:
> @doronbehar It's already called with it so you just have to remove with qt5 and add those all as args.

OK I'll update `qstopmotion`.

Reply:
> > The callPackage function doesn't actually bring any extra dependencies. It just brings those attributes into scope. (and ensures correct versions)
> 
> I understand but I don't understand how would you expect (in guvcview) that `stdenv.mkDerivation` will be used if `useGtk == true` vs `libsForQt5.mkDerivation` if `useQt == true`. I guess that's possible using a `let ... in` where the `stdenv` will be chosen but I don't understand what would be the benefit considering the price of readability and that only 1 `qt5` package is used there.


@doronbehar I believe you've misread libsForQt5.**callPackage** for libsForQt5.**mkDerivation**, which doesn't exist.
You simply need to:

```diff
- guvcview = callPackage
+ guvcview = libsForQt5.callPackage
```

to do the first change suggested here.

Reply:
Oh I think I get you now, we'll use `libsForQt5.callPackage` although not necessarily even a single `qt5.` dependency is even added to the inputs... BUT, won't that add `wrapQtAppsHook` unconditionally? It shouldn't be added if `useGtk == true;` so I guess I can do:

```
dontWrapQtApps = !useGtk
```

I haven't tested that yet but I'm afraid it will change the `makeWrapperArgs` or something. If it'll work I don't mind but I don't see how making these changes will make such a big difference.

Reply:
> Oh I think I get you now, we'll use libsForQt5.callPackage although not necessarily even a single qt5. dependency is even added to the inputs... 

I don't understand that, the diff I marked here adds `qtbase` to buildInputs incorrectly by referencing `qt5.*`.

> BUT, won't that add wrapQtAppsHook unconditionally? 

Nope, as mentioned the `mkDerivaiton` does this, not the `callPackage` function. I haven't suggested to use qt5's mkDerivation here.

Reply:
>  If it'll work I don't mind but I don't see how making these changes will make such a big difference.

The difference is in the nixpkgs manual

> Add a Qt application to all-packages.nix using libsForQt5.callPackage instead of the usual callPackage. The former ensures that all dependencies are built with the same version of Qt. 

Unfortunately we're intersecting things here, so it applies even though it's not strictly a Qt App.

Reply:
Oh right, I definitely got confused between the differences of `stdenv.mkderivation` and qt's `mkDerivation` vs the differences between `libsForQt5.callPackage` and the regular `callPackage`. So to sum up:

We'll use `stdenv.mkDerivation` while the wrap hooks will be chosen as before but `libsForQt5.callPacakge` will be used unconditionally so when `useQt` is set, mixed qt5 versions won't be used. The later change has no influence on wrapping.

Reply:
Yep, that's right @doronbehar :+1:

Reply:
Now both of them should line up to your request.